src/core/ext/transport/chaotic_good/tcp_frame_transport.cc

```diff
@@ -108,34 +98,39 @@ auto TcpFrameTransport::WriteFrame(const FrameInterface& frame) {
       // ... then write it to the control endpoint
       [this, &header, &frame]() {
         SliceBuffer output;
-        TcpFrameHeader{header, 0}.Serialize(
-            output.AddTiny(TcpFrameHeader::kFrameHeaderSize));
+        TcpFrameHeader hdr{header, 0};
+        GRPC_TRACE_LOG(chaotic_good, INFO)
+            << "CHAOTIC_GOOD: Send control frame " << hdr.ToString();
+        hdr.Serialize(output.AddTiny(TcpFrameHeader::kFrameHeaderSize));
         frame.SerializePayload(output);
         return control_endpoint_.Write(std::move(output));
       },
       // ... otherwise write it to a data connection
       [this, header, &frame]() mutable {
-        SliceBuffer payload;
-        const size_t padding = DataConnectionPadding(header.payload_length,
-                                                     options_.encode_alignment);
-        frame.SerializePayload(payload);
+        SliceBuffer control_bytes;
+        SliceBuffer data_bytes;
+        auto tag = next_payload_tag_;
+        ++next_payload_tag_;
+        TcpFrameHeader hdr{header, tag};
         GRPC_TRACE_LOG(chaotic_good, INFO)
-            << "CHAOTIC_GOOD: Send " << payload.Length()
+            << "CHAOTIC_GOOD: Send control frame " << hdr.ToString();
+        hdr.Serialize(control_bytes.AddTiny(TcpFrameHeader::kFrameHeaderSize));
+        const size_t padding = DataConnectionPadding(
+            TcpFrameHeader::kFrameHeaderSize + header.payload_length,
+            options_.encode_alignment);
+        frame.SerializePayload(data_bytes);
+        GRPC_TRACE_LOG(chaotic_good, INFO)
+            << "CHAOTIC_GOOD: Send " << data_bytes.Length()
             << "b payload on data channel; add " << padding << " bytes for "
             << options_.encode_alignment << " alignment";
         if (padding != 0) {
           auto slice = MutableSlice::CreateUninitialized(padding);
           memset(slice.data(), 0, padding);
-          payload.AppendIndexed(Slice(std::move(slice)));
+          data_bytes.AppendIndexed(Slice(std::move(slice)));
         }
-        return Seq(
-            data_endpoints_.Write(std::move(payload)),
-            [this, header](uint32_t connection_id) mutable {
-              SliceBuffer header_frame;
-              TcpFrameHeader{header, connection_id + 1}.Serialize(
-                  header_frame.AddTiny(TcpFrameHeader::kFrameHeaderSize));
-              return control_endpoint_.Write(std::move(header_frame));
-            });
+        return DiscardResult(
```

Comment:
Why are we swapping the order of control and data endpoint writes ?

Reply:
We used to have to do the data endpoint write to get the load balancing decision... now we don't, so we can do a join between them.

Ordering is pretty arbitrary, if it matters we can toy with it once we start benchmarking this work.